mask/eolearn/mask/cloud_mask.py

```diff
@@ -389,96 +380,105 @@ def _do_single_temporal_cloud_detection(self, bands):
+            if local_avg is None or (min_tdx, max_tdx) != (prev_min_tdx, prev_max_tdx):
+                local_avg, local_var = self._update_batches(local_avg, local_var, bands_slice, is_data_slice, sigma)
 
             # Interweave and concatenate
             multi_features = self._extract_multi_features(
-                bands_t, is_data_t, loc_mu, loc_var, rel_t_i, masked_bands, sigma
+                bands_slice, is_data_slice, local_avg, local_var, rel_tdx, masked_bands, sigma
             )
 
-            multi_proba[t_i * img_size : (t_i + 1) * img_size] = self._run_prediction(
+            multi_proba[tdx * img_size : (tdx + 1) * img_size] = self._run_prediction(
                 self.multi_classifier, multi_features
             )
 
-            prev_nt_min = nt_min
-            prev_nt_max = nt_max
+            prev_min_tdx = min_tdx
+            prev_max_tdx = max_tdx
 
-        return multi_proba[..., np.newaxis]
+        return multi_proba[..., None]
 
-    def _update_batches(self, loc_mu, loc_var, bands_t, is_data_t, sigma):
-        """Updates window variance and mean values for a batch"""
-        # Add window averages and variances to local data
-        if loc_mu is None:
-            win_avg_bands = self._map_sequence(bands_t, partial(self._win_avg, sigma=sigma))
-            win_avg_is_data = self._map_sequence(is_data_t, partial(self._win_avg, sigma=sigma))
-
-            win_avg_is_data[win_avg_is_data == 0.0] = 1.0
-
-            loc_mu = win_avg_bands / win_avg_is_data
-
-            win_prevars = self._map_sequence(bands_t, partial(self._win_prevar, sigma=sigma))
-            win_prevars -= loc_mu * loc_mu
+    def _update_batches(
+        self, local_avg: np.ndarray, local_var: np.ndarray, bands: np.ndarray, is_data: np.ndarray, sigma: float
+    ) -> Tuple[np.ndarray, np.ndarray]:
+        """
+        Calculates or updates the window average and variance.
+        The calculation is done per 2D image along the temporal and band axes.
+        """
+        local_avg_func = partial(self._win_avg, sigma=sigma)
+        local_var_func = partial(self._win_prevar, sigma=sigma)
 
-            loc_var = win_prevars
+        # take full batch if avg/var don't exist, otherwise take only last index slice
+        data = bands if local_avg is None else bands[-1][None, ...]
+        data_mask = is_data if local_avg is None else is_data[-1][None, ...]
 
-        else:
-            win_avg_bands = self._map_sequence(bands_t[-1][None, ...], partial(self._win_avg, sigma=sigma))
-            win_avg_is_data = self._map_sequence(is_data_t[-1][None, ...], partial(self._win_avg, sigma=sigma))
+        avg_data = self._map_sequence(data, local_avg_func)
+        avg_data_mask = self._map_sequence(data_mask, local_avg_func)
+        avg_data_mask[avg_data_mask == 0.0] = 1.0
 
-            win_avg_is_data[win_avg_is_data == 0.0] = 1.0
+        var_data = self._map_sequence(data, local_var_func)
```

Comment:
This part confuses me. I don't understand why you would need to divide with `avg_data_mask`. IIRC, this is a value between 0 and 1 for each time and for each band. In case the value is 0, it's manually set to 1 (because of the division error).

I could see this as a sort of weighting, but I'm not sure why you would increase the values of `avg_data` by dividing with something smaller than 1.

Additionally, why isn't the same applied for the `var_data` ?

My intuition tells me that this should be dropped, but it's risky at this point...

Reply:
we'll probably need to do some heavier refactoring before this becomes a bit clearer.